plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -0,0 +1,274 @@
+                                    )
+                                    declarative_artifact.artifact = new_artifact
+                            pb.done = pb.done + len(content.d_artifacts)
+                            outstanding_downloads = outstanding_downloads - len(content.d_artifacts)
+                            await out_q.put(content)
+                else:
+                    if shutdown:
+                        break
+
+                if outstanding_downloads < self.max_concurrent_downloads:
+                    saturated = False
+        await out_q.put(None)
+
+
+async def artifact_saver(in_q, out_q):
+    """
+    Stages API stage that saves any unsaved :attr:`DeclarativeArtifact.artifact` objects.
+
+    This stage expects :class:`~pulpcore.plugin.stages.DeclarativeContent` units from `in_q` and
+    inspects their associated :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` object stores one
+    :class:`~pulpcore.plugin.models.Artifact`.
+
+    Any unsaved :class:`~pulpcore.plugin.models.Artifact` objects are saved. Each
+    :class:`~pulpcore.plugin.stages.DeclarativeContent` is sent to `out_q` after all of its
+    :class:`~pulpcore.plugin.stages.DeclarativeArtifact` objects have been handled.
+
+    This stage drains all available items from `in_q` and batches everything into one large call to
+    the db for efficiency.
+
+    Args:
+        in_q (:class:`asyncio.Queue`): The queue to receive
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        out_q (:class:`asyncio.Queue`): The queue to put
+            :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+
+    Returns:
+        The artifact_saver stage as a coroutine to be included in a pipeline.
+    """
+    shutdown = False
+    declarative_content_list = []
+    while not shutdown:
+        try:
+            content = in_q.get_nowait()
+        except asyncio.QueueEmpty:
+            if not declarative_content_list:
+                content = await in_q.get()
+                declarative_content_list.append(content)
+                continue
+        else:
+            declarative_content_list.append(content)
+            continue
+
+        artifacts_to_save = []
+        for declarative_content in declarative_content_list:
+            if declarative_content is None:
+                shutdown = True
+                break
+            for declarative_artifact in declarative_content.d_artifacts:
+                if declarative_artifact.artifact.pk is None:
+                    src_path = str(declarative_artifact.artifact.file)
+                    dst_path = declarative_artifact.artifact.storage_path(None)
+                    with open(src_path, mode='rb') as input_file:
+                        django_file_obj = File(input_file)
+                        default_storage.save(dst_path, django_file_obj)
```

Comment:
I don't think there is anything in the django `Storage` interface that requires the implementation be stateless and thus no guarantee.  Using the global singleton `default_storage` could be dangerous.  Recommend using (instantiating) `DefaultStorage` instead.

Reply:
I think this is a supported approach. I got it from the django docs here. https://docs.djangoproject.com/en/2.0/topics/files/#storage-objects

Reply:
Both approaches are supported but I believe constructing the `DefaultStorage` is safer for the reason given.

Reply:
I don't think I understand how it's safer or what danger we are preventing with that safety. Can you explain it to me some more?

Reply:
The `Storage` interface does not explicitly stipulate that implementations need to be safe to used as a global singleton. A stateful implementation *could* have unpredictable side effects when reused.  I'm mainly suggesting that since you have a choice, use the safer of the two.

IMHO, providing `default_storage` was a poor decision by the django project.

Reply:
I'm going to switch this to use `DefaultStorage()` instead which will ensure each instance is an unrelated copy. https://docs.djangoproject.com/en/2.0/ref/files/storage/#django.core.files.storage.DefaultStorage